internal/rbd/controllerserver.go

```diff
@@ -1694,20 +1694,265 @@ func (cs *ControllerServer) ControllerPublishVolume(
 		return nil, status.Error(codes.InvalidArgument, "Volume Capabilities cannot be empty")
 	}
 
-	return &csi.ControllerPublishVolumeResponse{
-		// the dummy response carry an empty map in its response.
-		PublishContext: map[string]string{},
-	}, nil
+	secrets := req.GetSecrets()
+
+	if acquired := cs.VolumeLocks.TryAcquire(req.GetVolumeId()); !acquired {
+		return nil, status.Errorf(codes.Aborted, util.VolumeOperationAlreadyExistsFmt, req.GetVolumeId())
+	}
+	defer cs.VolumeLocks.Release(req.GetVolumeId())
+	// If the CSI driver is not running on Kubernetes, we do not need to
+	// unfence the node, so we return immediately.
+	// This is because the unfencing logic is specific to Kubernetes environments.
+	if !k8s.RunsOnKubernetes() {
+		return &csi.ControllerPublishVolumeResponse{}, nil
+	}
+
+	err := cs.unfenceNode(ctx, req.GetNodeId(), req.GetVolumeId(), secrets, cs.Driver.IsSetAddressMetadataEnabled())
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to unfence node %s: %v", req.GetNodeId(), err)
+	}
+
+	return &csi.ControllerPublishVolumeResponse{}, nil
 }
 
-// ControllerUnPublishVolume is a dummy unpublish implementation to mimic a successful attach operation being a NOOP.
+// ControllerUnpublishVolume implements the CSI ControllerUnpublishVolume RPC.
 func (cs *ControllerServer) ControllerUnpublishVolume(
 	ctx context.Context,
 	req *csi.ControllerUnpublishVolumeRequest,
 ) (*csi.ControllerUnpublishVolumeResponse, error) {
 	if req.GetVolumeId() == "" {
 		return nil, status.Error(codes.InvalidArgument, "Volume ID cannot be empty")
 	}
+	if req.GetNodeId() == "" {
+		return nil, status.Error(codes.InvalidArgument, "Node ID cannot be empty")
+	}
```

Comment:
NodeId is optional the RPC implementation, we should remove it for all nodeID's (this is for future)

Reply:
I have added a comment for cephfs for this one. if NodeID is empty lets return success and dont fencing anything as we dont know what node need to be fenced